Added drift (mu) to GaussianRandomWalk.

akuz opened the pull request:
This change would allow specifying drift (mu) for a GaussianRandomWalk. So, now the value for (or a prior over) can be defined over mu as well. Note, the default behaviour is unchanged as the default value for mu is 0.0. Also, I deleted self.mean = 0.0 assignment, because it doesn't look like it's used anywhere.

akuz committed:
Added drift (mu) to GaussianRandomWalk.

twiecki commented on pymc3/distributions/timeseries.py:
```diff
@@ -41,19 +41,24 @@ class GaussianRandomWalk(Continuous):
         self.tau = tau
         self.sd = sd
         self.init = init
-        self.mean = 0.
```
Actually we add these for the case when you need to get summary statistics.

twiecki commented on pymc3/distributions/timeseries.py:
```diff
@@ -41,19 +41,24 @@ class GaussianRandomWalk(Continuous):
 
     Parameters
     ----------
+    mu: tensor
```
`mu : tensor`

twiecki commented on pymc3/distributions/timeseries.py:
```diff
@@ -41,19 +41,24 @@ class GaussianRandomWalk(Continuous):
     Parameters
     ----------
+    mu: tensor
+        innovation drift
```
Add note on defaulting to 0.

akuz committed:
Note on mu defaulting to 0, and returned self.mean = 0.

twiecki commented:
Have you tried whether this works? I don't think it should. The random walk is defined as:
`x_t ~ N(x_t-1, s)`

So the mean is already the previous value where you're also setting it to be some other value.

akuz commented:
See for example here about Gaussian Walk with Drift: 

http://people.brandeis.edu/~pmherb/MatlabBootCamp/simulatingdata.html

I am pretty sure about it ;) Just google "Gaussian Walk with Drift"

I will prepare another pull request with updates you requested

akuz commented:
I'll also send you a notebook with a test

akuz commented:
You're right, I am passing mu twice into Normal logp, should have been different, will submit update

akuz committed:
Fix for the drift in GaussianRandomWalk (tested).

akuz commented:
I updated the pull request, and tested. Please see attached notebook where I added a drift "m" to the stochastic volatility example, and it was inferred approximately correctly (around 0.002).

[pymc3_grw_drift.pdf](https://github.com/pymc-devs/pymc3/files/68240/pymc3_grw_drift.pdf)

twiecki commented:
Yes, that seems correct. I think calling it `mu` is slightly confusing though. Perhaps `drift`?

akuz commented:
Thomas, I will change the name to drift after work (arrived to work now, from where I can't contribute to github), and will let you know. P.S. Do you know what check is failing in Travis? Seems to be unrelated?

twiecki commented:
```
======================================================================
ERROR: pymc3.tests.test_examples.test_examples2('pymc3.examples.stochastic_volatility',)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda2/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_examples.py", line 15, in check_example
    example.run("short")
  File "/home/travis/build/pymc-devs/pymc3/pymc3/examples/stochastic_volatility.py", line 93, in run
    step = NUTS(model.vars, scaling=start, gamma=.25)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/step_methods/nuts.py", line 78, in __init__
    self.potential = quad_potential(scaling, is_cov, as_cov=False)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/step_methods/quadpotential.py", line 35, in quad_potential
    partial_check_positive_definite(C)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/step_methods/quadpotential.py", line 54, in partial_check_positive_definite
    i, = np.nonzero(np.logical_or(np.isnan(d), d <= 0))
FloatingPointError: invalid value encountered in less_equal
```

It means the stochastic volatility example is failing so that's definitely related.

akuz commented:
I will have a look after work. Could you please let me know how to run tests locally?

twiecki commented:
`nosetests pymc3/tests`

akuz commented:
Thanks, will let you know

akuz committed:
Renamed the GaussianRandomWalk mu parameter to drift,
and also moved it to the end of parameter list.

akuz commented:
Hi Thomas, the problem was that I put the new "mu" parameter as the first argument of GaussianRandomWalk class, and in the stochastic volatility example it was initialised with the wrong variable. I've moved the new parameter to the end, and renamed it to "drift". The travis tests should all pass now. Cheers, A

akuz commented:
Tests passed. Although I would prefer the new parameter to be called "mu" instead of "drift". For example, the parameters tau (or sd) go into the Normal innovation with the unchanged names, then why should "mu" be "drift". But, anything is good for me. Feel free to rename or not. Cheers

twiecki commented:
You do have a point in regards to `mu`.

akuz commented:
Shall I change back to mu? :) Without making it the first parameter...

twiecki commented:
Yes.

akuz committed:
Renamed GaussianRandomWalk parameter from drift to mu.

akuz commented:
Done.

twiecki commented:
Thanks @akuz!

akuz commented:
Cool, it's working pretty nicely. I am hoping to contribute a class and example on Gaussian process smoothing.

![smoothing](https://cloud.githubusercontent.com/assets/2409854/11962703/c47e3096-a8d9-11e5-98b3-32db392a3614.png)

twiecki commented:
@akuz That'd be a great contribution!
